src/lib/client.ts

```diff
@@ -256,40 +254,50 @@ export default class BACnetClient extends TypedEventEmitter<BACnetClientEvents>
 		return id - 1
 	}
 
-	private _invokeCallback(id: number, err: Error | null, result?: any): void {
-		const callback = this._invokeStore[id]
-		if (callback) {
+	private _resolveDeferred(id: number, err: Error, result?: undefined): void
+	private _resolveDeferred(
+		id: number,
+		err: null | undefined,
+		result: NetworkOpResult,
+	): void
+	private _resolveDeferred(
+		id: number,
+		err: Error | null | undefined,
+		result?: NetworkOpResult,
+	): void {
+		const deferred = this._invokeStore[id]
+		if (deferred) {
 			trace(`InvokeId ${id} found -> call callback`)
-			return void callback(err, result)
-		}
-		debug('InvokeId', id, 'not found -> drop package')
-		trace(`Stored invokeId: ${Object.keys(this._invokeStore)}`)
-	}
-
-	private _addCallback(id: number, callback: DataCallback): void {
-		const toCall: DataCallback = (err, data) => {
 			delete this._invokeStore[id]
-			clearTimeout(timeout)
-
 			if (err) {
-				debug(`InvokeId ${id} callback called with error:`, err)
+				deferred.reject(err)
 			} else {
-				trace(`InvokeId ${id} callback called with data:`, data)
+				deferred.resolve(result)
 			}
-
-			callback(err, data)
+		} else {
+			debug('InvokeId', id, 'not found -> drop package')
+			trace(`Stored invokeId: ${Object.keys(this._invokeStore)}`)
 		}
+	}
 
+	private _addDeferred(id: number): Promise<NetworkOpResult> {
+		const deferred = new Deferred<NetworkOpResult>()
+		this._invokeStore[id] = deferred
+		trace(
+			`InvokeId ${id} callback added -> timeout set to ${this._settings.apduTimeout}.`, // Stack: ${new Error().stack}`,
+		)
 		const timeout = setTimeout(
-			toCall.bind(this, new Error('ERR_TIMEOUT')),
+			deferred.reject.bind(null, new Error('ERR_TIMEOUT')),
```

Comment:
when timeout triggers could it happen that for some reason the deferred promise resolves? Do we have a guard for this? I usually override `resolve/reject` functions to handle this

Reply:
I don't _think_ this is possible as of the latest commit that refactored everything into the dedicated `InvokeStore` class but I'm also not 100% sure. Will check later in the day.